### PR TITLE
AG-13280 - update sparkline interfaces

### PR DESCRIPTION
**Pull request by StephM-W**

### CODE REVIEW

File: packages/ag-charts-community/src/api/preset/sparkline.ts [hunk 1]
```diff
@@ -235,15 +242,27 @@ function axisPreset(
...
 }
 
-function gridLinePreset(opts: AgSparklineAxisOptions | undefined, defaultEnabled: boolean): AgAxisGridLineOptions {
+function gridLinePreset(
+    opts: AgSparklineAxisOptions | undefined,
+    defaultEnabled: boolean,
+    sparkOpts: AgSparklineOptions
+): AgAxisGridLineOptions {
     const gridLineOpts: AgAxisGridLineOptions = {};
+    const isBarVertical = sparkOpts.type === 'bar' && sparkOpts.direction !== 'horizontal';
```
Comment (jacobp100) [inline packages/ag-charts-community/src/api/preset/sparkline.ts]: We have a check for `isVertical` (called `swapAxes`) below. Maybe we could pass that boolean in instead

File: packages/ag-charts-community/src/api/preset/sparkline.ts [hunk 2]
```diff
@@ -293,6 +312,10 @@ export function sparkline(opts: AgSparklineOptions): AgCartesianChartOptions {
     const { data, series: [seriesOverrides] = [] } = sparklineDataPreset(baseData);
 
     const seriesOptions = optsRest as any as AgCartesianSeriesOptions;
+
+    if (seriesOptions.highlightStyle?.item && seriesOptions.type !== 'bar') {
+    }
+
```
Comment (jacobp100) [inline packages/ag-charts-community/src/api/preset/sparkline.ts]:
```suggestion
```

File: packages/ag-charts-types/src/chart/labelOptions.ts [hunk 3]
```diff
@@ -15,7 +15,7 @@ export interface AgChartLabelOptions<TDatum, TParams> extends Toggleable, FontOp
     formatter?: Formatter<AgChartLabelFormatterParams<TDatum> & TParams>;
 }
 
-export interface AgChartLabelFormatterParams<TDatum> extends AgChartCallbackParams<TDatum> {
+export interface AgChartLabelFormatterParams<TDatum = any> extends AgChartCallbackParams<TDatum> {
```
Comment (jacobp100) [inline packages/ag-charts-types/src/chart/labelOptions.ts]: Preferably we don't do this so we remember to type them. Why was this needed?

File: packages/ag-charts-community/src/api/preset/sparkline.ts [hunk 4]
```diff
@@ -182,18 +186,21 @@ export function sparklineDataPreset(data: any[] | undefined): {
             const mappedData = data.map((y, x) => ({ x, y }));
             return { data: mappedData, series: [{ xKey: 'x', yKey: 'y' }] };
         } else if (Array.isArray(firstItem)) {
-            const mappedData = data.map(([x, y]) => ({ x, y }));
+            const mappedData = data.filter((item) => !!item).map(([x, y]) => ({ x, y }));
```
Comment (alantreadway) [inline packages/ag-charts-community/src/api/preset/sparkline.ts]: We shouldn't be doing this filtering here - what is the motivation for this?

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts [hunk 5]
```diff
@@ -28,17 +28,17 @@ export interface AgSparklineContinuousAxisOptions<TDatum extends Date | number>
 }
 
 export interface AgSparklineCategoryAxisOptions extends AgSparklineBaseCategoryAxisOptions {
-    type: 'category';
+    type?: 'category';
```
Comment (jacobp100) [inline packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]:
```suggestion
    type: 'category';
```

We should enforce it at the type level, but if they don't pass it in, we can just let it slide

File: packages/ag-charts-types/src/chart/labelOptions.ts (hunk 3 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/chart/labelOptions.ts]: The grid callbacks don't force users to type them, this is to align with the current convention

File: packages/ag-charts-community/src/api/preset/sparkline.ts [hunk 6]
```diff
@@ -235,15 +242,27 @@ function axisPreset(
...
         gridLineOpts.enabled ??= true;
+    } else if (isBarVertical) {
+        gridLineOpts.style = [{ stroke: DEFAULT_AXIS_GRID_COLOUR }];
+        gridLineOpts.enabled ??= true;
     }
     if (opts?.strokeWidth != null) {
         gridLineOpts.width = opts?.strokeWidth;
         gridLineOpts.enabled ??= true;
+    } else if (isBarVertical) {
+        gridLineOpts.width = 2;
```
Comment (alantreadway) [inline packages/ag-charts-community/src/api/preset/sparkline.ts]: We should avoid hardcoding where possible, why do we need a vertical-case specific hardcoding too? I'd normally suggest where feasible to put this fallback default in the theme.

File: packages/ag-charts-types/src/chart/labelOptions.ts (hunk 3 above)
Comment (alantreadway) [inline packages/ag-charts-types/src/chart/labelOptions.ts]: Is this needed?

File: packages/ag-charts-types/src/chart/tooltipOptions.ts [hunk 7]
```diff
@@ -61,7 +61,7 @@ export interface AgTooltipRendererResult {
     class?: CssColor;
 }
 
-export interface AgSeriesTooltipRendererParams<TDatum> extends AgChartCallbackParams<TDatum> {
+export interface AgSeriesTooltipRendererParams<TDatum = any> extends AgChartCallbackParams<TDatum> {
```
Comment (alantreadway) [inline packages/ag-charts-types/src/chart/tooltipOptions.ts]: Ditto?

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 1 above)
Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts]: that other check isn't quite the same, it checks for !bar || vertical. In this case, this must only apply to vertical bar charts (we don't currently default to showing the gridline for line or area charts)

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts (hunk 5 above)
Comment (alantreadway) [inline packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]: I'd prefer we kept these required, what was the motivation here?

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts [hunk 8]
```diff
@@ -28,17 +28,17 @@ export interface AgSparklineContinuousAxisOptions<TDatum extends Date | number>
...
 }
 
 export interface AgSparklineNumberAxisOptions
     extends AgSparklineBaseAxisOptions,
         AgSparklineContinuousAxisOptions<number> {
-    type: 'number';
+    type?: 'number';
```
Comment (alantreadway) [inline packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]: Ditto

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts [hunk 9]
```diff
@@ -28,17 +28,17 @@ export interface AgSparklineContinuousAxisOptions<TDatum extends Date | number>
...
 }
 
 export interface AgSparklineTimeAxisOptions extends AgSparklineBaseAxisOptions, AgSparklineContinuousAxisOptions<Date> {
-    type: 'time';
+    type?: 'time';
```
Comment (alantreadway) [inline packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]: Ditto?

File: packages/ag-charts-types/src/series/cartesian/barOptions.ts [hunk 10]
```diff
@@ -22,7 +22,7 @@ export interface AgBarSeriesLabelOptions<TDatum, TParams> extends AgChartLabelOp
     padding?: PixelSize;
 }
 
-export interface AgBarSeriesItemStylerParams<TDatum>
+export interface AgBarSeriesItemStylerParams<TDatum = any>
```
Comment (alantreadway) [inline packages/ag-charts-types/src/series/cartesian/barOptions.ts]: Ditto

File: packages/ag-charts-community/src/api/preset/sparkline.ts [hunk 11]
```diff
@@ -182,18 +186,21 @@ export function sparklineDataPreset(data: any[] | undefined): {
...
 }
 
 function axisPreset(
     opts: AgSparklineAxisOptions | undefined,
     defaultType: AgCartesianAxisOptions['type']
 ): AgCartesianAxisOptions {
+    opts = {
+        ...opts,
+        type: opts?.type ?? defaultType,
+    } as any as AgSparklineAxisOptions;
```
Comment (alantreadway) [inline packages/ag-charts-community/src/api/preset/sparkline.ts]: Presuming was can avoid this if `axis.type` is required?

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts (hunk 5 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]: Updated with a different approach, let me know what you think

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 4 above)
Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts]: Our current specs state that we ignore undefined entries in the array - happy to move the filtering elsewhere. Discovered it when one of our examples threw an NPE here

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 6 above)
Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts]: It's only applicable for vertical bar charts, and our themes don't allow that level of specificity. But, as the default for horizontal bar charts is hidden anyway, we can move this default to the theme

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 11 above)
Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts]: It currently defaults to 'category' in prod, so it's optional from the user perspective. There's a comment higher up where we're discussing al alternative approach

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts (hunk 5 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]: Ok, my proposed doesn't fit the bill for other reasons - could you suggest a way to achieve the strong typing optionality you mention?

File: packages/ag-charts-types/src/chart/labelOptions.ts (hunk 3 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/chart/labelOptions.ts]: As above, to align with the grid convention for callbacks that can optionally be untyped

File: packages/ag-charts-types/src/chart/tooltipOptions.ts (hunk 7 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/chart/tooltipOptions.ts]: As above, to align with the grid convention for callbacks that can optionally be untyped

File: packages/ag-charts-types/src/series/cartesian/barOptions.ts (hunk 10 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/series/cartesian/barOptions.ts]: As above, to align with the grid convention for callbacks that can optionally be untyped

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 1 above)
Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts]: Simplified

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 6 above)
Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts]: Moved to theme

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 11 above)
Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts]: Removed, type is now required

File: packages/ag-charts-types/src/series/cartesian/barOptions.ts (hunk 10 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/series/cartesian/barOptions.ts]: Reverted

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts (hunk 9 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]: Reverted

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts (hunk 8 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]: Reverted

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts (hunk 5 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]: Reverted

File: packages/ag-charts-types/src/chart/tooltipOptions.ts (hunk 7 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/chart/tooltipOptions.ts]: Reverted

File: packages/ag-charts-types/src/chart/labelOptions.ts (hunk 3 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/chart/labelOptions.ts]: Reverted

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 4 above)
Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts]: Moving to the grid's sparklineRenderer

File: packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts (hunk 5 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/presets/sparkline/sparklineAxisOptions.ts]: Reverted

File: packages/ag-charts-types/src/chart/labelOptions.ts (hunk 3 above)
Comment (StephM-W) [reply packages/ag-charts-types/src/chart/labelOptions.ts]: Reverted

File: packages/ag-charts-community/src/api/preset/sparkline.ts [hunk 12]
```diff
@@ -245,6 +275,9 @@ function gridLinePreset(opts: AgSparklineAxisOptions | undefined, defaultEnabled
         gridLineOpts.width = opts?.strokeWidth;
         gridLineOpts.enabled ??= true;
     }
+    if (sparkOpts.type === 'bar' && sparkOpts.direction !== 'horizontal') {
+        gridLineOpts.enabled ??= true;
+    }
```
Comment (jacobp100) [inline packages/ag-charts-community/src/api/preset/sparkline.ts:277]: Could we just pass in `direction` here since we're working it out below?

**Review by jacobp100 — APPROVED** [review]

File: packages/ag-charts-community/src/api/preset/sparkline.ts [hunk 13]
```diff
@@ -67,6 +68,26 @@ const crossHairTooltip = {
...
+    },
+    log: {
+        gridLine: barGridLineDefaults,
+    },
+    time: {
+        gridLine: barGridLineDefaults,
+    },
+    category: {
+        gridLine: barGridLineDefaults,
+    },
```
Comment (jacobp100) [inline packages/ag-charts-community/src/api/preset/sparkline.ts:85]: Do we need `ordinal-time` too?

Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts:85]: Actually, current spec is only category, number and time. I'll remove log.

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 12 above)
Comment (StephM-W) [reply packages/ag-charts-community/src/api/preset/sparkline.ts:277]: I don't think the change is trivial, especially when direction really only applies to bar sparklines here.

**Review by alantreadway — APPROVED** [review]
LGTM
